File: vdm_pc/download/m3u8.py
"""M3U8 解析（移植自擴充 lib/m3u8.js）。"""
from __future__ import annotations

from dataclasses import dataclass
from urllib.parse import urljoin, urlparse


@dataclass
class Segment:
    url: str
    duration: float = 0.0


@dataclass
class Playlist:
    is_variant: bool
    playlists: list[dict]
    segments: list[Segment]


def resolve_url(base: str, ref: str) -> str:
    return urljoin(base, ref)


def parse_m3u8(text: str, base_url: str) -> Playlist:
    lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    playlists: list[dict] = []
    segments: list[Segment] = []
    bandwidth = 0
    resolution = 0

    i = 0
    while i < len(lines):
        line = lines[i]
        if line.startswith("#EXT-X-STREAM-INF"):
            bw = _match_int(line, r"BANDWIDTH=(\d+)")
            bandwidth = bw or 0
            res = _match_int(line, r"RESOLUTION=\d+x(\d+)")
            resolution = res or 0
            nxt = lines[i + 1] if i + 1 < len(lines) else ""
            if nxt and not nxt.startswith("#"):
                playlists.append(
                    {
                        "url": resolve_url(base_url, nxt),
                        "bandwidth": bandwidth,
                        "resolution": resolution,
                    }
                )
                i += 1
        elif line.startswith("#EXTINF"):
            dur = _match_float(line, r"#EXTINF:([\d.]+)") or 0.0
            nxt = lines[i + 1] if i + 1 < len(lines) else ""
            if nxt and not nxt.startswith("#"):
                segments.append(Segment(url=resolve_url(base_url, nxt), duration=dur))
                i += 1
        i += 1

    return Playlist(is_variant=bool(playlists), playlists=playlists, segments=segments)


def pick_best_variant(playlist: Playlist) -> str:
    if not playlist.playlists:
        return ""
    best = max(playlist.playlists, key=lambda p: (p.get("resolution") or 0, p.get("bandwidth") or 0))
    return str(best.get("url") or "")


def _match_int(text: str, pattern: str) -> int | None:
    import re

    m = re.search(pattern, text, re.I)
    return int(m.group(1)) if m else None


def _match_float(text: str, pattern: str) -> float | None:
    import re

    m = re.search(pattern, text, re.I)
    return float(m.group(1)) if m else None


def is_video_url(url: str) -> bool:
    from vdm_pc.browser.detector import is_likely_ad_url, is_video_url as _detect

    return _detect(url) and not is_likely_ad_url(url)

File: extension-src/lib/m3u8.js
(function () {
  const VDM = self.VDM;

  VDM.parseM3u8 = (text, baseUrl) => {
    const lines = text.split(/\r?\n/).map((l) => l.trim()).filter(Boolean);
    const playlists = [];
    const segments = [];
    let bandwidth = 0;
    let resolution = 0;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      if (line.startsWith("#EXT-X-STREAM-INF")) {
        const bw = line.match(/BANDWIDTH=(\d+)/i);
        bandwidth = bw ? parseInt(bw[1], 10) : 0;
        const res = line.match(/RESOLUTION=\d+x(\d+)/i);
        resolution = res ? parseInt(res[1], 10) : 0;
        const next = lines[i + 1];
        if (next && !next.startsWith("#")) {
          playlists.push({
            url: VDM.resolveUrl(baseUrl, next),
            bandwidth,
            resolution,
          });
          i++;
        }
      } else if (line.startsWith("#EXTINF")) {
        const durMatch = line.match(/#EXTINF:([\d.]+)/);
        const duration = durMatch ? parseFloat(durMatch[1]) : 0;
        const next = lines[i + 1];
        if (next && !next.startsWith("#")) {
          segments.push({
            url: VDM.resolveUrl(baseUrl, next),
            duration,
          });
          i++;
        }
      }
    }

    return {
      isVariant: playlists.length > 0,
      playlists,
      segments,
      targetDuration: 0,
    };
  };

  VDM.m3u8Meta = (playlist, bandwidth = 0) => {
    let duration = playlist.segments.reduce((s, seg) => s + (seg.duration || 0), 0);
    let estimatedSize = 0;
    if (bandwidth && duration) estimatedSize = Math.floor((bandwidth * duration) / 8);
    return { duration, estimatedSize };
  };

  VDM.fetchM3u8 = async (url, video) => {
    const referer = VDM.bestReferer(url, video.pageUrl, video.referer);
    video.referer = referer;
    VDM.log("info", "開始解析 M3U8", `${VDM.shortUrl(url)}\nReferer: ${referer.slice(0, 80)}`);

    if (video.tabId) {
      try {
        const text = await VDM.fetchTextInPage(video.tabId, url, video);
        const parsed = VDM.parseM3u8(text, url);
        VDM.log(
          "info",
          "M3U8 解析完成",
          parsed.isVariant
            ? `主清單，${parsed.playlists.length} 條變體`
            : `${parsed.segments.length} 個片段`
        );
        return parsed;
      } catch (err) {
        VDM.log("warn", "分頁內 M3U8 失敗，嘗試背景", err.message || String(err));
      }
    } else {
      VDM.log("warn", "無 tabId", "請在影片分頁開啟擴充後下載");
    }

    const headers = await VDM.buildHeaders(video, url, { forBackground: true });
    try {
      const res = await fetch(url, { headers });
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      const text = await res.text();
      return VDM.parseM3u8(text, url);
    } catch (err) {
      const msg = err.message || String(err);
      VDM.log("error", "M3U8 下載失敗", msg);
      throw new Error(
        video.tabId
          ? `M3U8 失敗：${msg}（請確認影片分頁仍開啟且正在播放）`
          : `M3U8 失敗：${msg}`
      );
    }
  };
})();
